test/main_x86.c

```diff
@@ -0,0 +1,325 @@
+    pthread_create (tid, NULL, (void*)&loop, NULL);
+}
+
+void cleanup (pthread_t tid) {
+#ifdef DEBUG
+    puts ("cleanup ()");
+#endif
+    enable = 0;
+    pthread_join (tid, NULL);
+    cleanup_router_regs ();
+    cleanup_routing_table ();
+}
+
+// TODO: replace while with something that won't hang up indefinitely
+uint32_t positive_test (uint32_t sport, uint32_t addr, uint32_t tport) {
+    uint32_t ret = 0;
+#ifdef DEBUG
+    printf ("positive_test (%d, %d, %d)\n", sport, addr, tport);
+#endif
+    grant[sport] = 0;
+    control[tport] = 0;
+    taddrs[sport] = addr;
+    requests[sport] = 1;
+    available[tport] = 1;
+    while (grant[sport] == 0);
+    requests[sport] = 0;
+    available[tport] = 0;
+    if (control[tport] != 1 << sport) {
+        ret = 1;
+        printf (
+            "\nsport=%d, addr=%d, tport=%d, expected %d got %d",
+            sport, addr, tport, 1<<sport, control[tport]
+        );
+        exit(1);
+    }
+    return ret;
+}
+
+// TODO: replace while with something that won't hang up indefinitely
+uint32_t negative_test (uint32_t sport, uint32_t addr) {
+    discard[sport] = 0;
+    taddrs[sport] = addr;
+    requests[sport] = 1;
+    while (discard[sport] == 0);
+    return 0;
+}
+
+uint32_t test (uint32_t pn, uint32_t lan, uint32_t on) {
+#ifdef DEBUG
+    printf ("test (%d, %d)\n", pn, lan);
+#endif
+    uint32_t i;
+    uint32_t j;
+    uint32_t k;
+    uint32_t x;
+    uint32_t sport;
+    uint32_t addr;
+    uint32_t tport;
+    uint32_t ret = 0;
+    for (i=0;i<=lan;i++) {
+        for (j=0;j<pn;j++) {
+            for (k=0;k<pn;k++) {
+                for (x=0;x<on;x++) {
+                    sport = k;
+                    tport = j+x;
+                    if (tport >= pn) {
+                        tport = 0;
+                    }
+                    if (i == 0) {
+                        addr = j;
+                    } else {
+                        addr = 32+(i-1)*pn+j;
+                    }
+#ifdef DEBUG
+                    printf (
+                         "params pn=%d, lan=%d, i=%d, j=%d, k=%d, sport=%d, addr=%d, tport=%d\n", 
+                         pn, lan, i, j, k, sport, addr, tport
+                    );
+#endif
+                    ret |= positive_test (sport, addr, tport);
+                }
+            }
+        }
+    }
+    return ret;
+}
+
+uint32_t test_case (uint32_t pn, uint32_t lan, uint32_t on) {
+#ifdef DEBUG
+    printf ("test_case (pn=%d, lan=%d, on=%d) ", pn, lan, on);
+#endif
+    uint32_t ret;
+    pthread_t tid;
+    setup (pn, lan, on, &tid);
+    ret = test (pn, lan, on);
+#ifdef DEBUG
+    if (ret == 0) {
+        puts ("passed");
+    } else {
+        puts ("failed");
+    }
+#endif
+    cleanup (tid);
+    return ret;
+}
+
+uint32_t test_routing () {
+#ifdef DEBUG
+    puts ("test_routing");
+#endif
+    uint32_t pn;
+    uint32_t lan;
+    uint32_t on;
+    uint32_t cnt = 1;
+    uint32_t ret = 0;
+    for (pn=MINPORTS;pn<=MAXPORTS;pn++) {
```

Comment:
i suggest spaces around `=`, and operators. Add a space after each `;` in the for-loop.